compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td

```diff
@@ -61,17 +61,16 @@ def FLOW_DispatchWorkgroupsOp : FLOW_PureOp<"dispatch.workgroups", [
     and even several different variants for a particular target device
     (differing workgroup sizes, etc).
 
-    Because of the general nature of the op in this dialect the workgroup count
-    provided to the `flow.dispatch.workgroups` op is in an abstract untiled
-    domain. Unlike when lowering to the HAL dialect the number of dimensions is
-    unbounded and does not yet have the workgroup size factored into it. As the
-    dispatch is lowered the workgroup count range will be converted into a 3D
-    XYZ grid space and divided up by the workgroup size chosen for particular
-    target devices.
+    Because at this point in the layering devices have not yet been selected the
+    workgroup count cannot be fully evaluated. Instead workload parameters are
+    captured that are then passed to a function that when later evaluated
+    computes the actual workgroup count based on target information. The
+    workload is not limited to the 3D XYZ grid dispatch of the workgroup count
+    and can contain any number of parameters used to compute it.
```

Comment:
> This initial work is meant to incrementally break some of the existing assumptions around workload being 3D and entirely defined by the backends.

Cool! So could we support backends with 1D/2D/5D/etc. dispatches? The 3D grid dispatch has always felt pretty GPU specific to me, though I'd been coming around to appreciate it on CPUs as well (with different levels of caches). What sorts of backends/devices would benefit from other dispatch approaches?

Reply:
The HAL is still going to use 3D dispatch - all this does is allow the inputs to the function that compute the 3D grid parameters be variadic. 3D is the right balance for dispatch-level distribution (beyond just practicalities of that being what GPUs use): if you need more dimensions they are either better done at coarser levels (distribution to other queues/devices/nodes) or finer levels (tiling within workgroups down to subgroups and invocations) - neither of which will show up on the dispatch ops.

The assertion is that at the precise level of the dispatch interface there's no need for 20D spaces as they can't possibly carry any actionable semantic information. When you see people try to tunnel through things like 6D spaces it's usually just a layering violation where someone shoved a high level concept ("this fat ML op operates on 6D tensors") all the way through the system ("because we need to call this library function that requires all 6D even though it just computes 1-3 numbers from it internally"). We can avoid tunneling all that through because we control both sides. We could still call out to library functions needing all 6D but that's not something the dispatch level sees: it's just some arguments we construct within the dispatch to then make the call. If you were doing that kind of stuff you'd just set your 3D grid to 1x1x1 and not let IREE do any distribution, or peel off a dimension or two that you want to use for multithreading/etc and leave the rest internal (what we do today), etc.

The important thing is that we have an actual layering that lets us reason about these things; if you want to distribute across nodes you do that before IREE, if you want to distribute across devices/queues you do that with streams, if you want to distribute across processing elements you do that with dispatches, and if you want to distribute across execution units you do that inside of the dispatches.

Reply:
(this is a good question and one that should be documented somewhere - not sure where though!)